src/main/java/org/embulk/input/zendesk/clients/ZendeskRestClient.java

```diff
@@ -0,0 +1,273 @@
+            }
+            else {
+                if (retryAfter > 0) {
+                    logger.warn("Reached API limitation, wait for at least '{}' '{}'", retryAfter, TimeUnit.SECONDS.name());
+                }
+                else if (status != 429) {
+                    logger.warn(String.format("'%s' temporally failure.", status));
+                }
+                return true;
+            }
+        }
+
+        //Won't retry for 4xx range errors except above. Almost they should be ConfigError e.g. 403 Forbidden
+        if (status / 100 == 4) {
+            throw new ConfigException("Status '" + status + "', message '" + message + "'");
+        }
+
+        logger.warn("Server returns unknown status code '" + status + "' message '" + message + "'");
+        return true;
+    }
+
+    private HttpRequestBase createGetRequest(String url, PluginTask task)
+    {
+        HttpGet request = new HttpGet(url);
+        ImmutableMap<String, String> headers = buildAuthHeader(task);
+        if (headers != null) {
+            headers.forEach(request::setHeader);
+        }
+        return request;
+    }
+
+    private ImmutableMap<String, String> buildAuthHeader(PluginTask task)
+    {
+        Builder<String, String> builder = new Builder<>();
+        builder.put(AUTHORIZATION, buildCredential(task));
+        addCommonHeader(builder, task);
+        return builder.build();
+    }
+
+    private String buildCredential(PluginTask task)
+    {
+        switch (task.getAuthenticationMethod()) {
+            case BASIC:
+                return "Basic " + ZendeskUtils.convertBase64(String.format("%s:%s", task.getUsername().get(), task.getPassword().get()));
+            case TOKEN:
+                return "Basic " + ZendeskUtils.convertBase64(String.format("%s/token:%s", task.getUsername().get(), task.getToken().get()));
+            case OAUTH:
+                return "Bearer " + task.getAccessToken().get();
+        }
+        return "";
+    }
+
+    private void addCommonHeader(final Builder<String, String> builder, PluginTask task)
+    {
+        task.getAppMarketPlaceIntegrationName().ifPresent(s -> builder.put(ZendeskConstants.Header.ZENDESK_MARKETPLACE_NAME, s));
+        task.getAppMarketPlaceAppId().ifPresent(s -> builder.put(ZendeskConstants.Header.ZENDESK_MARKETPLACE_APP_ID, s));
+        task.getAppMarketPlaceOrgId().ifPresent(s -> builder.put(ZendeskConstants.Header.ZENDESK_MARKETPLACE_ORGANIZATION_ID, s));
+
+        builder.put(CONTENT_TYPE, ZendeskConstants.Header.APPLICATION_JSON);
+    }
+
+    private RateLimiter getRateLimiter(final HttpResponse response)
+    {
+        if (rateLimiter == null) {
+            rateLimiter = initRateLimiter(response);
+        }
+        return rateLimiter;
+    }
+
+    private static synchronized RateLimiter initRateLimiter(final HttpResponse response)
+    {
+        String rateLimit = "";
```

Comment:
what if there is no `x-rate-limit`? it means `Double.parseDouble(rateLimit)` will always throw out exception because `""` (empty string) is obviously not able to parse.

Reply:
updated the message to make it more clear